File: clients/typescript/src/model/building.ts
import { BuildingType } from "./building-type";
import { Stream } from "../stream";

/**
 * A building
 */
export class Building {
    /**
     * Building's type
     */
    buildingType: BuildingType
    /**
     * Current health
     */
    health: number
    /**
     * Amount of work done for current task
     */
    workDone: number
    /**
     * Number of tasks finished since last tick
     */
    lastTickTasksDone: number

    constructor(buildingType: BuildingType, health: number, workDone: number, lastTickTasksDone: number) {
        this.buildingType = buildingType;
        this.health = health;
        this.workDone = workDone;
        this.lastTickTasksDone = lastTickTasksDone;
    }

    /**
     * Read Building from input stream
     */
    static async readFrom(stream: Stream): Promise<Building> {
        let buildingType;
        buildingType = await BuildingType.readFrom(stream);
        let health;
        health = await stream.readInt();
        let workDone;
        workDone = await stream.readInt();
        let lastTickTasksDone;
        lastTickTasksDone = await stream.readInt();
        return new Building(buildingType, health, workDone, lastTickTasksDone)
    }

    /**
     * Write Building to output stream
     */
    async writeTo(stream: Stream) {
        let buildingType = this.buildingType;
        await buildingType.writeTo(stream);
        let health = this.health;
        await stream.writeInt(health);
        let workDone = this.workDone;
        await stream.writeInt(workDone);
        let lastTickTasksDone = this.lastTickTasksDone;
        await stream.writeInt(lastTickTasksDone);
    }
}

File: clients/go/model/building.go
package model

import "fmt"
import "io"
import . "spb_ai_champ/stream"

// A building
type Building struct {
    // Building's type
    BuildingType BuildingType
    // Current health
    Health int32
    // Amount of work done for current task
    WorkDone int32
    // Number of tasks finished since last tick
    LastTickTasksDone int32
}

func NewBuilding(buildingType BuildingType, health int32, workDone int32, lastTickTasksDone int32) Building {
    return Building {
        BuildingType: buildingType,
        Health: health,
        WorkDone: workDone,
        LastTickTasksDone: lastTickTasksDone,
    }
}

// Read Building from reader
func ReadBuilding(reader io.Reader) Building {
    var buildingType BuildingType
    buildingType = ReadBuildingType(reader)
    var health int32
    health = ReadInt32(reader)
    var workDone int32
    workDone = ReadInt32(reader)
    var lastTickTasksDone int32
    lastTickTasksDone = ReadInt32(reader)
    return Building {
        BuildingType: buildingType,
        Health: health,
        WorkDone: workDone,
        LastTickTasksDone: lastTickTasksDone,
    }
}

// Write Building to writer
func (building Building) Write(writer io.Writer) {
    buildingType := building.BuildingType
    WriteInt32(writer, int32(buildingType))
    health := building.Health
    WriteInt32(writer, health)
    workDone := building.WorkDone
    WriteInt32(writer, workDone)
    lastTickTasksDone := building.LastTickTasksDone
    WriteInt32(writer, lastTickTasksDone)
}

// Get string representation of Building
func (building Building) String() string {
    stringResult := "{ "
    stringResult += "BuildingType: "
    buildingType := building.BuildingType
    stringResult += BuildingTypeToString(buildingType)
    stringResult += ", "
    stringResult += "Health: "
    health := building.Health
    stringResult += fmt.Sprint(health)
    stringResult += ", "
    stringResult += "WorkDone: "
    workDone := building.WorkDone
    stringResult += fmt.Sprint(workDone)
    stringResult += ", "
    stringResult += "LastTickTasksDone: "
    lastTickTasksDone := building.LastTickTasksDone
    stringResult += fmt.Sprint(lastTickTasksDone)
    stringResult += " }"
    return stringResult
}